---
title: "2023 â™¥ Poseidon"
author: "Clemens Schmid"
date: "2023-12-20"
categories: [reviews]
description: An opinionated summary of the year's events in and around the Poseidon framework.
image: "isba2023poster_media/house.png"
---

It's late December and the time of the year when work slows down around here. For many of us an opportunity to take a break and to look back, contemplating the achievements of the year. I decided to do so as well and write a bit about Poseidon.

What follows is a subjective account of the events in and around the framework in 2023 - each of my colleagues in the core team (Stephan Schiffels, Ayshin Ghalichi, Thiseas C. Lamnidis, Dhananjaya B. A. Mudiyanselage, Wolfgang Haak and I, Clemens Schmid) would probably emphasise different developments in such a write-up. That is in itself an achievement, because it shows how much the tech-stack, features and services in our little ecosystem have grown this year: beyond the understanding of each of us individually.

## The Poseidon schema

Let's start simple with the two new releases of the Poseidon schema we published this year: 2.7.0 and 2.7.1. They were published in short succession in March and May, the latter only slightly improving the .ssf files added in the first. See the changelog [here](https://www.poseidon-adna.org/#/changelog) for more details, but the addition of the [.ssf file](https://www.poseidon-adna.org/#/ssf_details) is indeed their most remarkable addition to the schema. With it we addressed a major desideratum and unresolved question in previous versions of Poseidon: How should genotype data be linked to the raw sequencing data on the ENA and other archives of the International Nucleotide Sequence Database Collaboration?

The .ssf file is, I would argue, a smart solution for this question. It specifies the very variables already used in the ENA database, allows for an extremely flexible, yet not arbitrary n:m connection between the entities in a Poseidon package and the raw data products and it can be generated semi-automatically for most of the data in our public archives. It can also be used to organize local data repositories independent of public databases. The .ssf file is finally the very foundation on top of which the amazing Minotaur workflow is built (see below).

Generally, both the fact that only two Poseidon schema releases were necessary this year and that we could treat them as non-breaking changes indicate that we reached a certain level of maturity and stability here. We certainly have ideas how to extend it further in the future, but at the moment I'm optimistic that we can maintain long-term backwards compatibility. The process in which we discussed, specified and later improved the .ssf file definition to only then build Minotaur on top of it was a very satisfying professional experience.

## The Minotaur workflow

The Minotaur workflow is a semi-automatic workflow to reproducibly process published sequencing data into Poseidon packages. Developing that became possible, because Thiseas joined the Poseidon core team in 2023. He came up with a sophisticated, yet open and transparent implementation of this process, where authors maintain agency over data processing parameters. A full write-up for the website is [in progress](https://github.com/poseidon-framework/poseidon-framework.github.io/pull/54), so here is the summary Thiseas prepared for our [poster at the ISBA conference](https://blog.poseidon-adna.org/posts/isba2023poster.html) (see below):

Community members can request new packages by submitting a build recipe as a Pull Request against a dedicated GitHub repository. This recipe is created from a Sequencing Source File (.ssf), describing the sequencing data for the package and where it can be downloaded. Using the recipe, the sequencing data gets processed through [nf-core/eager](https://nf-co.re/eager) on computational infrastructure of MPI-EVA, using a standardised, yet flexible, set of parameters. The generated genotypes, together with descriptive statistics of the sequencing data (Endogenous, Damage, Nr_SNPs, Contamination), are compiled into a Poseidon package, and made available to users in the Minotaur Archive (see below).

The Minotaur workflow is a timely addition to the Poseidon framework, providing a flexible solution to reframe legacy data in uniformly processed packages. 

## The public archives

## The software tools

## Outreach

## Outlook
